approval/internal/handler/approvalrequest/handler.go

```diff
@@ -31,27 +30,11 @@ type ApprovalRequestHandler struct {
 func (h *ApprovalRequestHandler) CreateOrUpdate(ctx context.Context, approvalReq *approvalv1.ApprovalRequest) error {
 	log := log.FromContext(ctx)
 
-	if approvalReq.Spec.State != approvalReq.Status.LastState {
-		contextutil.RecorderFromContextOrDie(ctx).Eventf(approvalReq,
-			"Normal", "Notification", "State changed from %s to %s", approvalReq.Status.LastState, approvalReq.Spec.State,
-		)
-
-		if approvalReq.Spec.State == approvalv1.ApprovalStateRejected {
-			var err error
-			approvalReq.Status.NotificationRef, err = util.SendNotification(ctx, approvalReq, approvalReq.GetNamespace(), string(approvalReq.Spec.State), &approvalReq.Spec.Resource, &approvalReq.Spec.Requester)
-			if err != nil {
-				return err
-			}
-		}
-	}
-
-	if approvalReq.GetGeneration() == 1 && approvalReq.Status.NotificationRef == nil {
-		var err error
-		namespace := approvalReq.Spec.Resource.Namespace // namespace of the owner of the to-be-subscribed resource (i.e. target namespace)
-		approvalReq.Status.NotificationRef, err = util.SendNotification(ctx, approvalReq, namespace, string(approvalReq.Spec.State), &approvalReq.Spec.Resource, &approvalReq.Spec.Requester)
-		if err != nil {
-			return err
-		}
+	// handle the notifications first
+	err := handleNotifications(ctx, approvalReq)
+	if err != nil {
+		// todo - decide if we want to fail here, or a failed notification is acceptable
```

Comment:
Similar comment: In my opinion should not fail. But perhaps we should discuss this in a separate meeting